inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/pooling_gpu_b_fs_zyx_fsv16_imad.cl

```diff
@@ -44,6 +55,167 @@ inline ACCUMULATOR_TYPE FUNC(apply_pooling)(ACCUMULATOR_TYPE tmp, ACCUMULATOR_TY
 #endif
 }
 
+#if GLOBAL_POOLING
+__attribute__((intel_reqd_sub_group_size(FEATURE_SLICE_SIZE)))
+__attribute__((reqd_work_group_size(1, LWS, 1)))
+KERNEL(pooling_gpu_b_fs_zyx_fsv16)(
+    const __global INPUT0_TYPE* input,
+    __global OUTPUT_TYPE* output
+#if HAS_FUSED_OPS_DECLS
+    , FUSED_OPS_DECLS
+#endif
+) {
+    const uint set_idx = (uint)get_global_id(1);
+#if OUTPUT_DIMS == 4
+    const uint y    = set_idx;
+    const uint z    = 0;
+#else
+    const uint y    = set_idx % INPUT0_SIZE_Y;
+    const uint z    = set_idx / INPUT0_SIZE_Y;
+#endif
+    const uint bf   = (uint)get_global_id(2);
+    const uint f    = (bf * FEATURE_SLICE_SIZE) % ALIGN_TO(INPUT0_FEATURE_NUM, FEATURE_SLICE_SIZE);
+    const uint b    = (bf * FEATURE_SLICE_SIZE) / ALIGN_TO(INPUT0_FEATURE_NUM, FEATURE_SLICE_SIZE);
+
+    const uint loc_id   = (uint)get_local_id(1);
+
+#if OUTPUT_FEATURE_NUM % FEATURE_SLICE_SIZE != 0
+    const bool last_in_f_group = (f == FEATURE_SLICE_SIZE * ((INPUT0_FEATURE_NUM - 1) / FEATURE_SLICE_SIZE));
+#endif
+
+    const int offset_x = -PADDING_SIZE_X;
+
+    ACCUMULATOR_VEC16 result = INIT_VAL;
+    __local ACCUMULATOR_VEC16 lg_storage[LWS];
+
+    for (uint pz = 0; pz < Z_LOAD; pz++) {
+        __attribute__((opencl_unroll_hint(Y_LOAD)))
+        for (uint py = 0; py < Y_LOAD; py++) {
+#if INPUT0_DIMS == 4
+            uint y_ind = y * Y_LOAD + py;
+            uint input_idx = INPUT0_GET_INDEX(b, f, y_ind, offset_x);
+            if (y_ind >= INPUT0_SIZE_Y)
+                continue;
+#else
+            uint z_ind = z * Z_LOAD + pz;
+            uint y_ind = y * Y_LOAD + py;
+            uint input_idx = INPUT0_GET_INDEX(b, f, z_ind, y_ind, offset_x);
+            if (z_ind >= INPUT0_SIZE_Z || y_ind >= INPUT0_SIZE_Y)
+                continue;
+#endif
+            for (uint px = 0; px < POOL_SIZE_X; px++) {
```

Comment:
In the other comment regarding unrolling I didn't mean that it should be disabled completely. You just need to limit the unrolling size. E.g. find on the host side max divider for POOL_SIZE_X in range [1:X] where X - is some reasonable value that won't lead to the instruction cache misses in the upper loops.